dex/networks/eth/txdata.go

```diff
@@ -10,40 +10,134 @@ import (
+// specific version of the swap contract. It returns the the list of redemptions
+// done in the call and errors if the call data does not call redeem with expected
+// argument types.
+func ParseRedeemData(calldata []byte, contractVersion uint32) ([]*Redemption, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.parseRedeemData(calldata)
+}
+
+// ParseRefundData parses the calldata used to call the refund function of a
+// specific version of the swap contract. It returns the secret hash and errors
+// if the call data does not call refund with expected argument types.
+func ParseRefundData(calldata []byte, contractVersion uint32) ([32]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return [32]byte{}, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.parseRefundData(calldata)
+}
 
-// ParseInitiateData accepts call data from a transaction that pays to a
-// contract with extra data. It will error if the call data does not call
-// initiate with expected argument types. It returns the array of initiations
-// with which initiate was called.
-func ParseInitiateData(calldata []byte) ([]swapv0.ETHSwapInitiation, error) {
+// PackInitiateData converts a list of Initiation to the call data for the
+// initiate function for the contract specified by contractVersion.
+func PackInitiateData(initiations []*Initiation, contractVersion uint32) ([]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.packInitiateData(initiations)
+}
+
+// PackRedeemData converts a list of Redemption to the call data for the redeem
+// function for the contract specified by contractVersion.
+func PackRedeemData(redemptions []*Redemption, contractVersion uint32) ([]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.packRedeemData(redemptions)
+}
+
+// PackRefundData converts a secret hash to the call data for the refund function
+// for the contract specified by contractVersion.
+func PackRefundData(secretHash [32]byte, contractVersion uint32) ([]byte, error) {
+	txDataHandler, ok := txDataHandlers[contractVersion]
+	if !ok {
+		return nil, fmt.Errorf("contract version %v does not exist", contractVersion)
+	}
+
+	return txDataHandler.packRefundData(secretHash)
+}
+
+type txDataHandler interface {
+	parseInitiateData([]byte) ([]*Initiation, error)
+	parseRedeemData([]byte) ([]*Redemption, error)
+	parseRefundData([]byte) ([32]byte, error)
+	packInitiateData([]*Initiation) ([]byte, error)
+	packRedeemData([]*Redemption) ([]byte, error)
+	packRefundData([32]byte) ([]byte, error)
```

Comment:
Usually we try to leave code that is exclusive to a test in the test file. Is there a good reason to define these here?

Reply:
good point.. `packInitiateData` is used in a test in the client/eth package, but I could just duplicate that in the other test.